Add triple state switch widget

vinothvino42 opened the pull request:
Ticket: #1154

`value = off | mixed | on`

Doc - https://github.com/EnsembleUI/ensemble_docs/pull/123

**Sample EDL** 

```yaml
View:
  header:
    title: Triple Switch
  body:
    Column:
      styles: { gap: 24, padding: 24, scrollable: true }
      children:
        - TripleSwitch:
            leadingText: "Entire Place "
            value: mixed
            enabled: true
            styles:
              {
                activeThumbColor: white,
                inactiveThumbColor: grey,
                activeColor: 0xFFD6D6CE,
                inactiveColor: 0xFF425C59,
                mixedColor: 0xFF858174,
              }
            onChange: |
              //@code
              console.log("Switch Value: "+this.value);

```

https://github.com/EnsembleUI/ensemble/assets/12869588/3c88ebde-4e0e-46d0-b222-f9e9f7c3c3ed

snehmehta commented on assets/schema/ensemble_schema.json:
```diff
@@ -1849,6 +1849,35 @@
       ]
     },
     "Switch-payload": {
+      "type": "object",
```
hey vinoth, the schema needs to add in ensemble-web-studio ts files, check the studio repo readme

vinothvino42 replied:
Okay I'll update in the studio once this PR is merged

snehmehta commented on lib/widget/triple_state_switch.dart:
```diff
@@ -0,0 +1,296 @@
+  void onToggle(SwitchState newValue) {
+    widget.onToggle(newValue);
+    widget._controller.value = newValue;
+    if (widget._controller.onChange != null) {
```
add a dispatch event here, so if any listener is bind to switch state change will update as well. 
Something like this 
https://github.com/EnsembleUI/ensemble/blob/afa3e8dcfa0eb0e6fbc16a783f03426b2b77b71f/lib/layout/tab_bar.dart#L160-L161

snehmehta replied:
Yes, my bad. It works. For some reason i thought the case of tabbar and footer not updating on tab change because of event but I was wrong it update I check my theory.

snehmehta commented on lib/widget/triple_state_switch.dart:
```diff
@@ -0,0 +1,296 @@
+      widget: FormField<bool>(
+        key: validatorKey,
+        validator: (value) {
+          if (widget._controller.required) {
```
do test this validation case, from the look of it won't work as expected

vusters commented on lib/widget/triple_state_switch.dart:
```diff
@@ -0,0 +1,296 @@
+  }
+
+  @override
+  Map<String, Function> setters() {
```
why copy and pasted all these code? 99% of it identical to the regular switch. You should at least use a base class instead of replicating all these.

vinothvino42 replied:
Refactored

vusters commented on lib/widget/triple_state_switch.dart:
```diff
@@ -0,0 +1,296 @@
+  Map<String, Function> setters() {
+    return {
+      'value': (value) => _controller.value =
+          SwitchState.values.from(Utils.getString(value, fallback: 'start')) ??
```
You have two "start" values here plus the default "start" - that is 3 different places to change if you decide to change the default state. Potential cause for bug unnecessary.
Use this: 
`_controller.value = SwitchState.values.from(value) ?? _controller.value`

vinothvino42 replied:
Done

vinothvino42 commented on lib/widget/triple_state_switch.dart:
```diff
@@ -0,0 +1,296 @@
+  }
+}
+
+enum SwitchState {
```
Changed it

vinothvino42 commented on lib/widget/triple_state_switch.dart:
```diff
@@ -0,0 +1,296 @@
+import 'package:ensemble_ts_interpreter/invokables/invokable.dart';
+import 'package:flutter/material.dart';
+
+class EnsembleTripleStateSwitch extends StatefulWidget
```
Now, I'm extending the FormFieldController for the TripleStateSwitchController. Currently, the FormFieldController is not migrated to the new way (extending the EnsembleWidgetController instead of WidgetController). I tried changing the FormFieldController so I am supposed to migrate all the widget which uses the FormFieldController to new way of building widgets (EnsembleWidget<Controller>).

So as of now I done in the old way only

vusters commented on lib/widget/switch.dart:
```diff
@@ -32,15 +42,16 @@ class EnsembleSwitch extends StatefulWidget
       'activeColor': (color) => _controller.activeColor = Utils.getColor(color),
       'inactiveColor': (color) =>
           _controller.inactiveColor = Utils.getColor(color),
+      'intermediateColor': (color) =>
```
mixedColor

vinothvino42 replied:
Changed it

vusters commented on lib/widget/switch.dart:
```diff
@@ -50,28 +61,30 @@ class EnsembleSwitch extends StatefulWidget
 
-  void onToggle(bool newValue) {
-    setProperty('value', newValue);
+  void setSwitchValue(dynamic value) {
```
Switch and TripleSwitch's value should be succinct. The former takes in true/false, the latter 3 states. There should be no guessing like this, so please refactor so their inputs are separate. The schema should reflect tis too

vinothvino42 replied:
Added `onToggle` in those widgets with their state as input parameter

vusters commented on lib/widget/switch.dart:
```diff
@@ -50,28 +61,30 @@ class EnsembleSwitch extends StatefulWidget
-  Color? inactiveThumbColor;
 
-  framework.EnsembleAction? onChange;
+  void onToggle(dynamic newValue) {
```
same thing here. Each widget should have its own onToggle

vinothvino42 replied:
Updated it

vusters commented on lib/widget/switch.dart:
```diff
@@ -102,12 +115,13 @@ class SwitchState extends FormFieldWidgetState<EnsembleSwitch> {
     return InputWrapper(
-      type: EnsembleSwitch.type,
+      // type: EnsembleTripleStateSwitch.type,
+      type: '',
```
put the type in. This is to help with exceptions.

vinothvino42 replied:
Done

vinothvino42 commented:
@vusters Can you review it? Just updated it, also checked the validation and it's working fine

- Moved the `onToggle` to Switch and TripleSwitch with their input as parameter and updated the logic in the SwitchBase class for value-changing
- Changed the widget name from `TripleStateSwitch` to `TripleSwitch`
- Added the `type` to the InputWrapper widget
- Changed the property from `intermediateColor` to `mixedColor`
